WKDownloadDelegate.decideDestinationUsingResponse indicating a file that already exists should cancel instead of crashing network process

achristensen07 opened the pull request:
#### 4827b5fe002be3e7b195492caeba1092bb0a03b3
<pre>
WKDownloadDelegate.decideDestinationUsingResponse indicating a file that already exists should cancel instead of crashing network process
<a href="https://bugs.webkit.org/show_bug.cgi?id=285316">https://bugs.webkit.org/show_bug.cgi?id=285316</a>
<a href="https://rdar.apple.com/142036600">rdar://142036600</a>

Reviewed by Richard Robinson.

When I introduced WKDownload I introduced some restrictions on decideDestinationUsingResponse
that were well intentioned but untested.  Richard found that if you improperly specify the destination,
we get an invalid IPC to a DownloadProxy object that no longer exists.  The didFail calls are
redundant.  Calling the completionHandler with an empty destination will make the download cancel.

* Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
(TestWebKitAPI::DestinationFileAlreadyExists)):

Canonical link: <a href="https://commits.webkit.org/288401@main">https://commits.webkit.org/288401@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cf238518c130d96b7ca46f5294c55f24b14cf57

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82983 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2626 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37276 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88094 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34031 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85075 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2701 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10546 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64642 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22399 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86038 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1981 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75465 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44920 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1888 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29659 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33065 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30348 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89460 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10267 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7438 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73059 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71275 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72279 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16451 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15196 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1646 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12843 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10220 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15732 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10092 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13557 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11860 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
```diff
@@ -131,20 +131,10 @@ void decideDestinationWithSuggestedFilename(WebKit::DownloadProxy& download, con
-                return;
-            }
+            if (![fileManager fileExistsAtPath:[destination URLByDeletingLastPathComponent].path])
+                return completionHandler(WebKit::AllowOverwrite::No, { });
```
is there a way of letting the client know what went wrong? because if they give a URL that they think is valid but is actually invalid, it will seem as if it got cancelled for no reason when in actuality it was cancelled because of one of these reasons. Maybe a release log or something?

rr-codes replied:
oh nvm, they are completely redundant

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
```diff
@@ -3145,4 +3145,33 @@ HTTPServer childFrameServer({
+TEST(WKDownload, DestinationFileAlreadyExists)
+{
+    HTTPServer server { { { "/"_s, { "hi"_s } } } };
+    auto webView = adoptNS([WKWebView new]);
```
these all should be `RetainPtr` instead of `auto` right per the smart pointer guidelines, right?

achristensen07 committed:
WKDownloadDelegate.decideDestinationUsingResponse indicating a file that already exists should cancel instead of crashing network process
https://bugs.webkit.org/show_bug.cgi?id=285316
rdar://142036600

Reviewed by Richard Robinson.

When I introduced WKDownload I introduced some restrictions on decideDestinationUsingResponse
that were well intentioned but untested.  Richard found that if you improperly specify the destination,
we get an invalid IPC to a DownloadProxy object that no longer exists.  The didFail calls are
redundant.  Calling the completionHandler with an empty destination will make the download cancel.

* Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm:
(TestWebKitAPI::DestinationFileAlreadyExists)):

Canonical link: https://commits.webkit.org/288401@main
